controllers/control/k8ssandratask_controller.go

```diff
@@ -0,0 +1,212 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package control
+
+import (
+	"context"
+	"fmt"
+	cassapi "github.com/k8ssandra/cass-operator/apis/control/v1alpha1"
+	k8sapi "github.com/k8ssandra/k8ssandra-operator/apis/k8ssandra/v1alpha1"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/clientcache"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/config"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/labels"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/utils"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	api "github.com/k8ssandra/k8ssandra-operator/apis/control/v1alpha1"
+)
+
+// K8ssandraTaskReconciler reconciles a K8ssandraTask object
+type K8ssandraTaskReconciler struct {
+	*config.ReconcilerConfig
+	client.Client
+	Scheme      *runtime.Scheme
+	ClientCache *clientcache.ClientCache
+}
+
+//+kubebuilder:rbac:groups=control.k8ssandra.io,namespace="k8ssandra",resources=k8ssandratasks,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=control.k8ssandra.io,namespace="k8ssandra",resources=k8ssandratasks/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=control.k8ssandra.io,namespace="k8ssandra",resources=k8ssandratasks/finalizers,verbs=update
+
+func (r *K8ssandraTaskReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithValues("K8ssandraTask", req.NamespacedName)
+
+	logger.Info("Fetching task", "K8ssandraTask", req.NamespacedName)
+	k8Task := &api.K8ssandraTask{}
+	err := r.Get(ctx, req.NamespacedName, k8Task)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{RequeueAfter: r.ReconcilerConfig.DefaultDelay}, err
+	}
+	// TODO k8Task.DeepCopy()? why?
+
+	if k8Task.DeletionTimestamp != nil {
+		// TODO delete dependent objects?
```

Comment:
The problem is that cass-operator already makes the `CassandraDatacenter` the owner of a `CassandraTask`. If I also make `K8ssandraTask` an owner, then both `K8ssandraTask` and `CassandraDatacenter` would have to be removed, which is not what we want.
I really think removing a `K8ssandraTask` should remove the associated `CassandraTasks` though, is there another way to achieve that?

Reply:
AFAIK ownership references do not work across namespaces. I know for certain that they do not work across multiple Kubernetes clusters. Since we can have CassandraTasks that could be in different namespaces as well as different k8s clusters compared to the "owning" K8ssandraTask, I think deletion logic here..

Reply:
Even in that case, the deletion logic should not delete CassandraTasks, let the cass-operator do that. Otherwise we'll have competing TTLs and those TTLs will cause incorrect and unnecessary reconciliations.

If deleting in the incorrect time, the CassandraDatacenter will reference to a task that's not suddenly there, causing .. unwanted behavior.

Reply:
I would also prefer to keep the deletion logic in a single operator. However the kind of scenario I'm concerned about is:
- the user creates a scheduled `K8ssandraTask`. k8ssandra-operator automatically creates a `CassandraTask` in each DC.
- the user changes their mind and deletes the `K8ssandraTask`. If we don't do anything in k8ssandra-operator, the `CassandraTasks` still exist.
- the user thinks they've cancelled the jobs, but they will still run

> If deleting in the incorrect time, the CassandraDatacenter will reference to a task that's not suddenly there, causing .. unwanted behavior.

But the reference goes the other way, right? AFAICT it's the task's labels that reference the `CassandraDatacenter`.

Also, isn't that the same as user-triggered deletion in a cass-operator-only scenario?

> Otherwise we'll have competing TTLs and those TTLs will cause incorrect and unnecessary reconciliations.

I haven't fully thought out TTL yet. But I think a good approach might be:
- when k8ssandra-operator creates the `CassandraTasks`, it doesn't set a TTL on them.
- instead, it sets a TTL on the `K8ssandraTask`, and has its own logic to manage it.
- once the `K8ssandraTask`'s TTL expires, it gets deleted, and all its `CassandraTasks` with it (assuming we've solved cascading deletion).

Reply:
I've pushed a provisional implementation of deletion and TTL based on the ideas above.